k2/csrc/cuda/debug.cuh

```diff
@@ -0,0 +1,169 @@
+            filename, line,
+            cudaGetErrorString(error));
+    fflush(stderr);
+  #elif (K2_PTX_ARCH >= 200)
+    printf("CUDA error ID=%d, NAME=%s, [block (%d,%d,%d) thread (%d,%d,%d), %s, %d]: %s\n",
+           error, cudaGetErrorName(error),
+           blockIdx.z, blockIdx.y, blockIdx.x, threadIdx.z, threadIdx.y, threadIdx.x,
+           filename, line,
+           cudaGetErrorString(error));
+  #endif
+    if (abort) {
+      exit(EXIT_FAILURE);
+    }
+  }
+#endif
+  return error;
+}
+
+/**
+ * \brief Macro for checking cuda error.
+ * If not cudaSuccess and K2_MAKE_ERROR_CHECK, print and exit.
+ */
+#ifndef K2_CUDA_CHECK_ERROR
+  #define K2_CUDA_CHECK_ERROR(e) \
+    k2::K2_CUDA_DEBUG((cudaError_t) (e), __FILE__, __LINE__, true)
+#endif
+
+/**
+ * \brief Macro for checking cuda standard runtime api return status.
+ * If api return status is not cudaSuccess, print and exit.
+ * Thus it's actually could be an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * \note:
+ * All runtime api return an error code, but for an asynchronous api,
+ * the error code only reports errors that occur on the host
+ * prior to executing the task, typically related to parameter validation.
+ * The device kernel error would unfortunately be left for next runtime api call.
+ * To check cuda runtime async api, as cuda kernel launches are asynchronous also,
+ * the same macro for kernel could be used.
+ * See `K2_CUDA_KERNEL_SAFE_CALL`.
+ *
+ * \usage:
+ * K2_CUDA_API_SAFE_CALL(cuda_runtime_api())
+ */
+#ifndef K2_CUDA_API_SAFE_CALL
+  #define K2_CUDA_API_SAFE_CALL(c) K2_CUDA_CHECK_ERROR(c)
+#endif
+
+/**
+ * \brief Macro for checking cuda standard runtime api return status.
+ * If api return status is not cudaSuccess, print and exit.
+ * Thus it's actually an alias of K2_CUDA_CHECK_ERROR.
+ *
+ * \note:
+ * Kernel launches do not return any error code, thus checking should after it.
+ * To wait kernel to finish, cudaDeviceSynchronize is called between.
+ * And to avoid pre-launch error disturbing, check error before kernel launch is
+ * also needed (but not necessary if one promise each error is take cared properly)
+ *
+ * \usage:
+ * K2_CUDA_KERNEL_SAFE_CALL(kernel_foo<<<>>>())
+ */
+#ifndef K2_CUDA_KERNEL_SAFE_CALL
+  #define K2_CUDA_KERNEL_SAFE_CALL(c)                 \
+    do {                                              \
+      K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+      K2_CUDA_API_SAFE_CALL(c);                       \
+      K2_CUDA_API_SAFE_CALL(cudaDeviceSynchronize()); \
+      K2_CUDA_CHECK_ERROR(cudaGetLastError());        \
+    } while (0)
+#endif
+
+/**
+ * \brief Log macro for printf statements.
+ * `printf` is supported by both host and device.
+ * This Log is for debugging, the error msg is printed to the stderr.
+ *
+ * Refer to:
+ * https://github.com/NVlabs/cub/blob/c3cceac115c072fb63df1836ff46d8c60d9eb304/cub/util_debug.cuh#L109
+ */
+#if !defined(K2_DLOG)
+  #if !(defined(__clang__) && defined(__CUDA__))
```

Comment:
The indentation looks weird.

Reply:
😂 Please give more details. As to me, the auto-format of my IDE had aligned the macros but it made it less readable i think.

Reply:
OK, I think I know what you mean. An indent error here is fixed now.